### PR TITLE
Fix GDB pretty-printer for tuples and pointers

### PR DESCRIPTION
**Pull request by gentoo90**
Names of children should not be the same, because GDB uses them to distinguish the children.

|Before|After|
|---|---|
|tuples_before|tuples_after|

`main.rs`
```rust
enum Test {
    Zero,
    One(i32),
    Two(i32, String),
    Three(i32, String, Vec<String>),
}

fn main() {
    let tuple = (1, 2, "Asdfgh");
    let zero = Test::Zero;
    let one = Test::One(10);
    let two = Test::Two(42, "Qwerty".to_owned());
    let three = Test::Three(9000,
                            "Zxcvbn".to_owned(),
                            vec!["lorem".to_owned(), "ipsum".to_owned(), "dolor".to_owned()]);
    println!(""); // breakpoint here
}
```

`launch.json`
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "gdb",
            "request": "launch",
            "gdbpath": "rust-gdb",
            "name": "Launch Program",
            "valuesFormatting": "prettyPrinters", //this requires plugin Native Debug >= 0.20.0
            "target": "./target/debug/test_pretty_printers",
            "cwd": "${workspaceRoot}"
        }
    ]
}
```

### CODE REVIEW

**Comment by carols10cents** [comment]
Thanks for the PR! We’ll periodically check in on it to make sure that @alexcrichton or someone else from the team reviews it soon.

**Comment by gentoo90** [comment]
Fixed exception in pointers parsing

|Before|After|
|---|---|
|ptr_before|ptr_after|

**Comment by alexcrichton** [comment]
r? @michaelwoerister 

( knows more about debuginfo and pretty printers than I )

This looks great though!

**Comment by gentoo90** [comment]
`str()` failed on things like `let opt_s = Some("I \u{2665} Unicode");`

**Comment by michaelwoerister** [comment]
Thanks a lot for the PR, @gentoo90! I would be great if you could add test cases to the `src/test/debuginfo` suite so those fixes don't break again in the future. Let me know if you need any help with that.

**Comment by michaelwoerister** [comment]
You could try `./x.py test --stage 0 src/test/debuginfo`. That should spare you as much compiling as possible.

**Comment by gentoo90** [comment]
Most of the tests are `ignored`, accept `gdb-pretty-struct-and-enums-pre-gdb-7-7.rs`. Is this OK? Is this because I don't have `lldb` installed?

**Comment by michaelwoerister** [comment]
You should not need LLDB for the GDB based tests. I assume this is on Linux? What's the GDB version you are using?

**Comment by gentoo90** [comment]
Yes, gentoo linux, gdb 7.12

**Comment by gentoo90** [comment]
I've found another issue while trying to debug `bootstrap`. When it's built as a part of cargo workspace, it doesn't have info about its source files. So when I set a breakpoint in `src/bootstrap/bin/main.rs`, `gdb` says `No source file named /home/user/rust/src/bootstrap/bin/main.rs.`

But if I remove `src/Cargo.toml` and build `bootstrap` with `cargo build`, the breakpoint works just fine.

**Comment by michaelwoerister** [comment]
> Most of the tests are ignored...

I just tested with GDB 7.12 on Ubuntu and tests are not ignored for me. If you remove all `*.stamp` files from `build/x86_64-unknown-linux-gnu/test/debuginfo`, are tests still ignored during the first run of `x.py test --stage 0 src/test/debuginfo`? I get 13 ignored tests in that case, which should be fine.

> I've found another issue while trying to debug `bootstrap`.

Would you mind opening an issue about that?

**Comment by gentoo90** [comment]
A proper way to test tuple field names would be

```rust
// gdbr-command:interpreter-exec mi2 "-enable-pretty-printing"
// gdbr-check:^done
// gdbr-command:interpreter-exec mi2 "-var-create noPadding8 @ noPadding8"
// gdbr-check:^done,name="noPadding8",numchild="2",value="{...}",type="(i8, u8)",thread-id="1",has_more="0"
// gdbr-command:interpreter-exec mi2 "-var-list-children noPadding8"
// gdbr-check:"^done,numchild="2",displayhint="array",children=[child={name="noPadding8.0",exp="0",numchild="0",value="-100",type="i8",thread-id="1"},child={name="noPadding8.1",exp="1",numchild="0",value="100",type="u8",thread-id="1"}],has_more="0"
```

but `-enable-pretty-printing` doesn't do anything in batch mode and pretty-printers are not used :(. Looks like a bug in GDB.

**Comment by gentoo90** [comment]
Added separate pretty-printer for empty structs, which prints the struct type instead of `{...}` in [variable objects](https://sourceware.org/gdb/onlinedocs/gdb/GDB_002fMI-Variable-Objects.html) mode.

|Before|After|
|---|---|
|empty_before|empty_after|

Testing against variable objects is impossible again without `-enable-pretty-printing`, and CLI interface has the tests already.

File: src/test/debuginfo/pretty-std.rs [hunk 1]
```diff
@@ -38,6 +38,9 @@
 // gdbg-check:$6 = None
 // gdbr-check:$6 = core::option::Option::None
 
+// gdb-command: print some_string
+// gdbr-check:$7 = Some = {"IAMA optional string!"}
```
Comment (michaelwoerister) [inline src/test/debuginfo/pretty-std.rs]: We should try making this `gdb-check` instead of just `gdbr-check`, so both kinds of GDB are tested.

**Comment by michaelwoerister** [comment]
Ping @gentoo90, regarding https://github.com/rust-lang/rust/pull/42278#discussion_r120337071. Could you make these regular `gdb-command`s and `gdb-check`s? Otherwise we'll only test versions of GDB that already support Rust natively, while the pretty printers should also work for older GDB versions.

**Comment by gentoo90** [comment]
Sorry for delay.

> older GDB versions

ahh, so that's why `gdbg-*` stuff isn't runned on my system.

**Comment by michaelwoerister** [comment]
> Sorry for delay.

No worries. Thanks for continuing to work on this!

> ahh, so that's why gdbg-* stuff isn't runned on my system.

Yes, that seems like a likely reason. I'm actually not sure what version of GDB is on our test infrastructure.

**Comment by michaelwoerister** [comment]
@bors r+

Let's give it a try.

**Comment by michaelwoerister** [comment]
:tada: 
Thanks, @gentoo90!

**Pull request merged**